Fix: Error catching in WebSockets

EugeneWilson opened the pull request:
This pull request addresses issue [#89](https://github.com/matsjla/league-connect/issues/89), where the websocket times out when the LCU is first starting. I fixed this by converting the do-while loop in `createWebSocketConnection` into a recursive function that resolves if successful, retries if within `maxRetries`, and rejects if it's an error other than 'ECONNREFUSED' or if it exceeds `maxRetries`.

Summary of changes made:
- Added the `__internalRetryCount` property to the `ConnectionOptions` interface to keep track of the number of retries attempted, and the `maxRetries` property to limit the number of retries. `maxRetries` defaults to 10, but users can modify it to -1 for infinite retries or a number of their choosing.
- Removed the `__internalMockFaultyConnection` and `__internalMockCallback` properties, as they are no longer needed for debugging purposes, since we are not looking for thrown errors.
- Made the `options` argument optional in the `createWebSocketConnection` function, as it was already effectively optional.
- Marked `pollInterval` as an optional property as it seems was intended.

The changes in commit https://github.com/matsjla/league-connect/pull/71/commits/cec79ea0cc78090cc2487af2b1917eccd84abe85 did not fix the error because WebSockets do not throw errors, but instead use an 'error' event.

I've tested these changes by creating a new LeagueWebSocket as the client is launching, and it worked correctly. Please let me know if you have any questions or concerns.

matsjla commented:
This is excellent - thank you so much for the contribution, and thank you for researching the flaw in the previous fix!

I've taken a view at your changes and they look great - again, thanks!

Do you think you could update the tests/websocket.test.ts file before I merge? It would be greatly appreciated.

EugeneWilson commented:
I have not used Jest before, so I'm sure the tests could be done better with mocks. However, I was not able to figure out how to. These tests work consistently, and I'll look into it further when I have extra time to see if I can pretty up the testing. 
Since testing uses a real LeagueWebSocket instance, the LCU client needs to be open at the time of testing. Also, it was necessary to wait for the socket to open before emitting the error or there would be an error trying to close the socket before it was open.
### Tests added:
- ``Handling ECONNREFUSED error no retries`` should test that ECONNREFUSED errors are handled correctly when there are no retries.
- ``Handling ECONNREFUSED error with retries`` should test that ECONNREFUSED errors are handled correctly when there are retries. It's recommended to run this test with three retries or fewer as it may take a long time to run with more retries.
- ``Handling an unknown error`` should test that unknown errors are rejected correctly.
- ``EndTestOpen`` should test an ECONNREFUSED error that resolves itself and will return the WebSocket. This is similar to the LCU starting up.

Since I had to add the ``__internalMockCallback`` and ``__internalMockFaultyConnection`` properties back in, I marked them as ``@internal`` and set the ``stripInternal`` option to true in the TypeScript config. This should stop users from seeing them come up while using the package from NPM, but is still visible to contributors using the TS files. I also added the @internal flag to the ``__internalRetryCount``.

In the function `createWebSocketConnection`, the parameter `options` has been set to have a default empty object when not provided. This is now done in the function declaration instead of later in the error handler.

Finally, I updated the README to have the options for ``createWebSocketConnection`` and I changed rift-explorer to LCU-Explorer since rift-explorer is no longer maintained.
### README changes:
- Added `LeagueWebSocket` to the table of contents.
- Added ``createWebSocketConnection`` options to the README.
- Updated the ``createWebSocketConnection`` example to use some of the new options.
- Changed rift-explorer to LCU-Explorer in the README. Also, the link to the new LCU-Explorer repository has been updated.

matsjla commented:
Sorry for being late, but this looks great! Thank you so much

matsjla commented:
I've tagged rc10 now, should be available on NPM shortly
